processing/src/main/java/org/apache/druid/segment/virtual/ListFilteredVirtualColumn.java

```diff
@@ -0,0 +1,193 @@
+import java.util.Set;
+
+/**
+ * {@link VirtualColumn} form of {@link ListFilteredDimensionSpec}, powered by
+ * {@link org.apache.druid.query.dimension.ForwardingFilteredDimensionSelector}
+ */
+public class ListFilteredVirtualColumn implements VirtualColumn
+{
+  private final String name;
+  private final DimensionSpec delegate;
+  private final Set<String> values;
+  private final boolean allowList;
+
+  @JsonCreator
+  public ListFilteredVirtualColumn(
+      @JsonProperty("name") String name,
+      @JsonProperty("delegate") DimensionSpec delegate,
+      @JsonProperty("values") Set<String> values,
+      @JsonProperty("isAllowList") @Nullable Boolean isAllowList
+  )
+  {
+    this.name = Preconditions.checkNotNull(name, "name");
+    this.delegate = delegate;
+    this.values = values;
+    this.allowList = isAllowList == null ? true : isAllowList.booleanValue();
+  }
+
+
+  @JsonProperty("name")
+  @Override
+  public String getOutputName()
+  {
+    return name;
+  }
+
+  @JsonProperty
+  public Set<String> getValues()
+  {
+    return values;
+  }
+
+  @JsonProperty("isAllowList")
+  public boolean isAllowList()
+  {
+    return allowList;
+  }
+
+  @JsonProperty
+  public DimensionSpec getDelegate()
+  {
+    return delegate;
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    CacheKeyBuilder builder = new CacheKeyBuilder(VirtualColumnCacheHelper.CACHE_TYPE_ID_LIST_FILTERED)
+        .appendString(name)
+        .appendCacheable(delegate)
+        .appendStringsIgnoringOrder(values)
+        .appendBoolean(allowList);
+    return builder.build();
+  }
+
+
+  @Override
+  public DimensionSelector makeDimensionSelector(
+      DimensionSpec dimensionSpec,
+      ColumnSelectorFactory factory
+  )
+  {
+    if (allowList) {
+      return ListFilteredDimensionSpec.filterAllowList(values, factory.makeDimensionSelector(delegate));
+    } else {
+      return ListFilteredDimensionSpec.filterDenyList(values, factory.makeDimensionSelector(delegate));
+    }
+  }
+
+  @Override
+  public ColumnValueSelector<?> makeColumnValueSelector(
+      String columnName,
+      ColumnSelectorFactory factory
+  )
+  {
+    return makeDimensionSelector(DefaultDimensionSpec.of(columnName), factory);
+  }
+
+  @Override
+  public ColumnCapabilities capabilities(String columnName)
+  {
+    return new ColumnCapabilitiesImpl().setType(delegate.getOutputType());
+  }
+
+  @Override
+  public ColumnCapabilities capabilities(ColumnInspector inspector, String columnName)
+  {
+    return inspector.getColumnCapabilities(delegate.getDimension());
+  }
+
+  @Override
+  public List<String> requiredColumns()
+  {
+    return Collections.singletonList(delegate.getDimension());
+  }
+
+  @Override
+  public boolean usesDotNotation()
+  {
+    return false;
+  }
+
+  @Override
+  public BitmapIndex getBitmapIndex(
+      String columnName,
+      ColumnSelector selector
+  )
+  {
+    return selector.getColumnHolder(delegate.getDimension()).getBitmapIndex();
```

Comment:
Doesn't this need to be decorated in some way, to avoid including the values we wanted to be filtered out?

Like, if someone filters on this virtual column and names a value that isn't included, I think we wouldn't want that filter to match. In SQL `WHERE MV_CONTAINS(MV_FILTER_ONLY(col, ['x']), 'y')` should not match anything, even if `col` contains `y` sometimes.

Reply:
Yes, good catch :+1:

There was another bug hiding this, which is that `VirtualColumns.getBitmapIndex` still calls `VirtualColumn.capabilities(columnName)` instead of `VirtualColumn.capabilities(inspector, columnName)`, and the former was not setting the flag on the artificial `ColumnCapabilities` it produces indicating that the column had bitmap indexes. So, it was always using a matcher instead of the bitmap index, which produced the correct behavior. We should probably at some point in the future figure out how to update `VirtualColumns.getBitmapIndex` to also have a `ColumnInspector` handy so it can use the better `capabilities` method, but I'll save that for another day.

I've added specialized `BitmapIndex` implementations to override the behavior of `getIndex` to only return index if the filter it matches the filter for allow-list/doesn't match the filter for deny-list, and tests (the SQL no matches test would fail by picking up matches without these implementations (after the `capabilities` bug was fixed).

Reply:
By the way, to set expectations around your example of `MV_CONTAINS(MV_FILTER_ONLY(col, ['x']), 'y')`, sql planning isn't cool enough yet and just makes a single expression filter that contains this whole expression, but that could probably be improved in the future. Other usages of this virtual column expression and filters do use the native indexes as expected, so it isn't always the case, just depends partially on the filters I think. 

I believe it is potentially related to virtual columns not being able to accept other virtual columns as input but didn't dig in too deep.

Reply:
Got it. The SQL behavior is fine for now. The important part (to me) is just that the native features work without bugs.